### PR TITLE
O3-3342 Add support for program states in form builder

### PR DESCRIPTION
**Pull request by CynthiaKamau**
## Requirements

- [x] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-docs.openmrs.org/docs/frontend-modules/contributing.en-US#contributing-guidelines) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://om.rs/o3ui).
- [ ] My work includes tests or is validated by existing tests.

## Summary

This PR introduces support for program states in form builder

## Screenshots

https://github.com/openmrs/openmrs-esm-form-builder/assets/19533785/a36be81e-e00d-45f3-9e7f-7c20f92ed91b

Screenshot 2024-05-31 at 10 23 58

### CODE REVIEW

**Review by hadijahkyampeire — DISMISSED** [review]
LGTM, thanks @CynthiaKamau

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 1]
```diff
@@ -580,6 +604,76 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
                   />
                 </RadioButtonGroup>
               ) : null}
+
+              {questionType === 'programState' && (
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:608]: CleanShot 2024-05-31 at 12  00 46@2x

Can we add some margin between the fields?

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 2]
```diff
@@ -580,6 +604,76 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                <>
+                  <div>
+                    <TextInput
+                      id="program"
+                      labelText="Add program uuid"
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:613]: Can we fetch these programs from the backend and render them in a ComboBox or similar? I wouldn't expect the user to know which UUIDs exist ahead of time.

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 3]
```diff
@@ -580,6 +604,76 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                    <FormLabel className={styles.label}>
+                      {t('searchForBackingProgramState', 'Search for a backing program state')}
+                    </FormLabel>
+                    {isLoadingProgramWorkflow && (
+                      <InlineLoading
+                        className={styles.loader}
+                        description={t('searchingProgramStates', 'Searching program states') + '...'}
+                      />
+                    )}
+                    {programWorkflowLookupError ? (
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:652]: Can we also add this logic to the Edit question modal so it's possible to change these values later on via the interactive builder?

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 4]
```diff
@@ -580,6 +618,95 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                        )}
+                        id="programWorkflowLookup"
+                        items={programWorkflows}
+                        itemToString={(item: ProgramWorkflow) => item?.concept?.display}
+                        onChange={({ selectedItem }) => {
+                          handleProgramWorkflowChange(selectedItem as ProgramWorkflow);
+                        }}
+                        placeholder={t('addProgramWorkflow', 'Add program workflow')}
+                        selectedItem={programWorkflowUuid}
+                        className={styles.programsDropdown}
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:668]:
```suggestion
											 selectedItem={programWorkflows.find((workflow) => workflow.uuid === programWorkflowUuid)}
                        titleText={t('programWorkflow', 'Program workflow')}
```

1. The `<Stack>` component will apply the appropriate spacing between the fields. We wouldn't need the additional class.
2. Setting `selectedItem` to `programWorkflowUuid` means we won't see any text rendered in the `ComboBox`. We need to find the workflow whose UUID matches `programWorkflowUuid`.

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 5]
```diff
@@ -580,6 +618,95 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                      />
+                    )}
+                  </div>
+                  {programWorkflowUuid && (
+                    <div>
+                      {isLoadingProgramStates && (
+                        <InlineLoading
+                          className={styles.loader}
+                          description={t('searchingProgramStates', 'Searching program states') + '...'}
+                        />
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:678]: Ditto here re: `SelectSkeleton`

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 6]
```diff
@@ -580,6 +618,95 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                          kind="error"
+                          lowContrast
+                          className={styles.error}
+                          title={t('errorFetchingProgramState', 'Error fetching program state')}
+                          subtitle={t('pleaseTryAgain', 'Please try again.')}
+                        />
+                      )}
+                      {programStates?.length > 0 && (
+                        <MultiSelect
+                          label={t('programState', 'Program state')}
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:691]:
```suggestion
                          titleText={t('programState', 'Program state')}
```

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 7]
```diff
@@ -580,6 +618,95 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                          id="programState"
+                          helperText={t('chooseProgramState', 'Choose a program state')}
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:693]: I think we can get rid of this helperText.

File: src/components/interactive-builder/editable-value.scss [hunk 8]
```diff
@@ -20,3 +20,7 @@
 .sectionLabel {
   @include type.type-style('heading-02');
 }
+
+.programsDropdown {
+  margin-top: 1rem;
+}
```
Comment (denniskigen) [inline src/components/interactive-builder/editable-value.scss:26]: Should be obviated by `Stack`. See above.

File: src/components/interactive-builder/question-modal.scss [hunk 9]
```diff
@@ -126,3 +126,7 @@
   color: colors.$red-60;
   margin-left: 0.125rem;
 }
+
+.programsDropdown {
+  margin-top: 1rem;
+}
```
Comment (denniskigen) [inline src/components/interactive-builder/question-modal.scss:132]: Ditto.

File: src/components/interactive-builder/edit-question-modal.component.tsx
Comment (denniskigen) [inline src/components/interactive-builder/edit-question-modal.component.tsx:1]: Ditto here for most of the suggestions above.

**Review by denniskigen — COMMENTED** [review]
It's close. Just a few more suggestions to get it across the line.

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 10]
```diff
@@ -675,6 +709,80 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                    {selectedProgram && (
+                      <ComboBox
+                        id="programWorkflowLookup"
+                        items={programWorkflows}
+                        itemToString={(item: ProgramWorkflow) => item?.concept?.display}
+                        onChange={({ selectedItem }: { selectedItem: ProgramWorkflow }) =>
+                          handleProgramWorkflowChange(selectedItem)
+                        }
+                        placeholder={t('addProgramWorkflow', 'Add program workflow')}
+                        selectedItem={programWorkflowUuid}
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:749]:
```suggestion
                        selectedItem={programWorkflows.find((workflow) => workflow.uuid === programWorkflowUuid)}
```

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 11]
```diff
@@ -675,6 +709,80 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                      )}
+                      {selectedProgramState?.map((answer) => (
+                        <Tag className={styles.tag} key={answer?.uuid} type={'blue'}>
+                          {answer?.concept?.display}
+                        </Tag>
+                      ))}
+                    </div>
+                  )}
+                </Stack>
+              )}
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:785]: This whole bit should move into the `<Stack>` too. Also, we shouldn't need the parent div.

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 12]
```diff
@@ -675,6 +709,80 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                          id="programState"
+                          items={programStates}
+                          itemToString={(item: ProgramState) => (item ? item?.concept?.display : '')}
+                          selectionFeedback="top-after-reopen"
+                          onChange={(data: ProgramStateData) => setSelectedProgramState(data.selectedItems)}
+                          selectedItems={selectedProgramState}
+                        />
+                      )}
+                      {selectedProgramState?.map((answer) => (
+                        <Tag className={styles.tag} key={answer?.uuid} type={'blue'}>
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:778]: The container element with the `Tag`s could do with some top margin.

File: src/components/interactive-builder/add-question-modal.component.tsx (hunk 10 above)
Comment (CynthiaKamau) [reply src/components/interactive-builder/add-question-modal.component.tsx:749]: I think the variable name is confusing, i will change it to programWorkflow, a user should chose a workflow from the list, this change is handled on edit view

File: src/components/interactive-builder/add-question-modal.component.tsx [hunk 13]
```diff
@@ -675,6 +709,78 @@ const AddQuestionModal: React.FC<AddQuestionModalProps> = ({
...
+                          {answer?.concept?.display}
+                        </Tag>
+                      ))}
+                    </div>
+                  )}
+                </Stack>
+              )}
             </Stack>
           </FormGroup>
```
Comment (denniskigen) [inline src/components/interactive-builder/add-question-modal.component.tsx:785]: Also, could we gate the Inline date and obs comment `RadioButtonGroup` behind a check for whether the question type is `obs` so they're not always shown when editing any question type?

```tsx
{questionToEdit.type === 'obs' && (
  // ...
}
```

**Comment by denniskigen** [comment]
Does editing work OK for you? I'm finding that changes are not getting propagated to the schema the way I'd expect them to.

**Comment by CynthiaKamau** [comment]
> Does editing work OK for you? I'm finding that changes are not getting propagated to the schema the way I'd expect them to.

This is how it looks : https://www.loom.com/share/ab45bf24a2bb48b7994ba9e12287311e?sid=df458d34-e9d3-445d-ba4b-321a509b6f4c

**Review by ODORA0 — APPROVED** [review]
LGTM

**Comment by denniskigen** [comment]
I've just tested things out on dev3 and it looks like there are still some issues with this:

https://github.com/openmrs/openmrs-esm-form-builder/assets/8509731/9b78d24c-dece-44f1-8d18-79fa6bf6a461

- The selected value for the `Program workflow` field is not displayed in the UI
- The selected program states are not rendered correctly. Their parent flex container should have its flex-direction set to `row` to spread the tags on the horizontal axis.

**Pull request merged**